addons/survey/models/survey_survey.py

```diff
@@ -792,19 +796,18 @@ def _get_session_most_voted_answers(self):
         current_user_input_lines = current_user_inputs.mapped('user_input_line_ids').filtered(lambda answer: answer.suggested_answer_id)
 
         # count the number of vote per answer
-        votes_by_answer = dict.fromkeys(current_user_input_lines.mapped('suggested_answer_id'), 0)
+        votes_by_answer = dict.fromkeys(current_user_input_lines.mapped(lambda answer: (answer.suggested_answer_id, answer.matrix_row_id)), 0)
         for answer in current_user_input_lines:
-            votes_by_answer[answer.suggested_answer_id] += 1
+            votes_by_answer[(answer.suggested_answer_id, answer.matrix_row_id)] += 1
 
         # extract most voted answer for each question
-        most_voted_answer_by_questions = dict.fromkeys(current_user_input_lines.mapped('question_id'))
-        for question in most_voted_answer_by_questions.keys():
-            for answer in votes_by_answer.keys():
-                if answer.question_id != question:
+        most_voted_answer_by_questions = dict.fromkeys(current_user_input_lines.mapped(lambda answer: (answer.question_id, answer.matrix_row_id)))
+        for (question, question_row), most_voted_answer in most_voted_answer_by_questions.items():
+            for (answer, answer_row), answer_votes in votes_by_answer.items():
+                if answer.question_id != question or answer_row != question_row:
                     continue
-                most_voted_answer = most_voted_answer_by_questions[question]
-                if not most_voted_answer or votes_by_answer[most_voted_answer] < votes_by_answer[answer]:
-                    most_voted_answer_by_questions[question] = answer
+                if not most_voted_answer or votes_by_answer[most_voted_answer] < answer_votes:
+                    most_voted_answer_by_questions[(question, question_row)] = (answer, answer_row)
```

Comment:
I'm wondering if splitting 'normal' and 'matrix' conditional question with if conditions won't be easier to read than using tuple.
I agree this is convenient to use tuple as row can be false or not, implied by the question type and we don't have to specify the type in the code to make it work.. but this feels to me that it makes all that pretty hard(er) to read and follow.

Reply:
Left it as is as agreed.

small explanation: in any case, we would need tuples to handle the 'matrix' questions (because answers are the same for each row of a matrix, so we need to know which row will trigger a conditional question, in addition to the answer - maybe there is another way, but this one seemed to be the one requiring the less changes in the code).  So splitting normal and matrix triggering questions would just duplicate code, with the difference that we will have the previous version and the version I added at the same time, and won't be that much less confusing if less confusing at all.